neuralpredictors/training/tracking.py

```diff
@@ -146,3 +154,369 @@ def asdict(self, time_key="time", make_copy=True):
+        current_log = self._gather_log(n_log, key, index=-1)
+        if tqdm_iterator:
+            tqdm_iterator.set_postfix(**current_log)
+        else:
+            logger.info(current_log)
+
+    def check_isfinite(self, log: Optional[Union[Mapping, Sequence]] = None) -> bool:
+        """
+        Checks if all entries in `log` or (normalized) `self.log` are finite.
+        Args:
+            log: dict that is recursively searched for infinite entries
+
+        Returns: True if all entries are finite.
+        """
+        if log is None:
+            log = self._normalize_log(self.log)
+        if isinstance(log, abc.Mapping):
+            for k, l in log.items():
+                if not self._check_isfinite(l):
+                    return False
+        else:
+            return np.isfinite(log).any()
+        return True
+
+    def get_objective(self, log: Optional[Union[Mapping, Sequence]] = None, key: Tuple[str, ...] = ()) -> np.array:
+        """
+        Get the value of the objective that corresponds to the key.
+        Args:
+            log: log to retrieve objective from
+            key: key to match with the log
+
+        Returns:
+            value array (across epochs and normalized) for a specific objective key
+        """
+        if log is None:
+            log = self._normalize_log(self.log)
+        if len(key) > 1:
+            return self.get_objective(log[key[0]], key[1:])
+        else:
+            return log[key[0]]
+
+    def get_current_objective(self, key: Tuple[str, ...]) -> float:
+        return self.get_objective(self._normalize_log(self.log), key)[-1]
+
+    def get_current_main_objective(self, key: Tuple[str, ...]) -> float:
+        combined_key = key + self.main_objective if isinstance(key, tuple) else (key,) + self.main_objective
+        return self.get_current_objective(combined_key)
+
+    def finalize(self) -> None:
+        """ After training, normalize the log and save the total time. """
+        self.time = np.array(self.time)
+        self.time -= self.time[0]
+        self.log = self._normalize_log(self.log)
+
+    def state_dict(self) -> Dict[str, any]:
+        """
+        Serializes this instance to a Python dictionary.
+
+        Returns:
+            :obj:`Dict[str, any]`: Dictionary of all the attributes that make up this configuration instance,
+        """
+        output = copy.deepcopy(self.__dict__)
+        return output
+
+    def load_state_dict(self, tracker_dict: Dict) -> None:
+        """
+        Loads given state_dict from another tracker.
+        Args:
+            tracker_dict: state that should override this tracker
+
+        """
+        self.main_objective = tracker_dict["main_objective"]
+        self.objectives = tracker_dict["objectives"]
+        self.log = tracker_dict["log"]
+        self.time = tracker_dict["time"]
+        self.epoch = tracker_dict["epoch"]
+
+    @classmethod
+    def from_dict(cls, tracker_dict: Dict) -> "AdvancedTracker":
```

Comment:
You can put `from __future__ import annotations` at the top of the file and then you don't need the quotes around the return type.